### PR TITLE
finish user service grpc

### PR DESCRIPTION
**Pull request by bombnp**
- Convert user service to grpc (from REST)
- Fix user migrations and models (+ seed data)
- Add createUser functionality
- Add InferCodeFromError function to infer grpc status codes from postgres errors

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,6 +5,7 @@ go 1.19
 require (
 	github.com/go-redis/redis/v9 v9.0.0-beta.2
 	github.com/gofiber/fiber/v2 v2.38.1
+	github.com/lib/pq v1.10.2
```
Comment (nonya123456) [inline go.mod]: go mod tidy

File: lib/postgres/models/user.go [hunk 2]
```diff
@@ -3,12 +3,17 @@ package models
...
-	UserRole  RoleEnum = "user"
+	AdminRole RoleEnum = "ADMIN"
+	UserRole  RoleEnum = "USER"
 )
 
 type User struct {
-	ID    string   `json:"id" gorm:"column:id;primaryKey"`
-	Role  RoleEnum `json:"role" gorm:"column:role;type:role;not null"`
-	Email string   `json:"email" gorm:"column:email;not null"`
+	Id       *string   `json:"id" gorm:"column:id;primaryKey;default:gen_random_uuid()"`
```
Comment (nonya123456) [inline lib/postgres/models/user.go]: Why is Id a pointer?

File: user/cmd/main.go [hunk 3]
```diff
@@ -24,20 +28,22 @@ func main() {
...
+	reflection.Register(grpcServer)
 
-	userGroup := app.Group("/api/user")
-	userGroup.Get("/me", handlers.GetCurrentUser)
+	// Register UserService server
+	libproto.RegisterUserServiceServer(grpcServer, server.NewServer(repo))
 
-	err = app.Listen(":" + conf.Server.Port)
+	// Serve
+	lis, err := net.Listen("tcp", ":"+conf.Server.Port)
```
Comment (saenyakorn) [inline user/cmd/main.go:45]: Please implement graceful shutdown

File: user/internal/repository/repository_user.go [hunk 4]
```diff
@@ -10,3 +12,11 @@ func (r *Repository) GetUserById(id string) (*models.User, error) {
 	}
 	return &user, nil
 }
+
+func (r *Repository) CreateUser(user *models.User) error {
```
Comment (saenyakorn) [inline user/internal/repository/repository_user.go:16]: Should you construct the new struct instead of using model.User? In this case, that mean, we can also specify id when creating user?

Comment (bombnp) [reply user/internal/repository/repository_user.go:16]: why new struct tho when it has the same fields

Comment (bombnp) [reply user/internal/repository/repository_user.go:16]: though i could allow specific ids, and default to random if not provided instead

Comment (saenyakorn) [reply user/internal/repository/repository_user.go:16]: Ok

Comment (saenyakorn) [reply user/internal/repository/repository_user.go:16]: I think that the service should not be able to define id by itself. For example, assume 1 service by 2 instance, the first instance randomly generate uuid as A and the second also randomly generate uuid as A as well. They call createUser at the same time. One of them might get the error due to PK constraint but if we use uuid from Postgres this case will not happen (It do only one transaction at the time).

TL;DR my point is omitting id from the parameter of createUser

File: lib/postgres/models/user.go (hunk 2 above)
Comment (bombnp) [reply lib/postgres/models/user.go]: oops, this was back i was trying to send it as NULL

fixed

File: go.mod (hunk 1 above)
Comment (bombnp) [reply go.mod]: fixed

File: user/internal/repository/repository_user.go (hunk 4 above)
Comment (bombnp) [reply user/internal/repository/repository_user.go:16]: oh, i thought u meant u want to specify id

ok

**Review by natTP — APPROVED** [review]
Other than already commented,LGTM

File: user/internal/repository/repository_user.go (hunk 4 above)
Comment (bombnp) [reply user/internal/repository/repository_user.go:16]: it's a little hard to create a new type from models.User without the Id part tho

But isn't it normal anyways? that problem only occurs if the code specifies the Id when creating user, and we're not doing that.

Other cases of gorm also use normal models, even though it has ID field on it.

File: user/cmd/main.go (hunk 3 above)
Comment (bombnp) [reply user/cmd/main.go:45]: done

**Review by saenyakorn — APPROVED** [review]

**Review by nonya123456 — APPROVED** [review]
